### PR TITLE
Add support for patching external packages

### PR DESCRIPTION
**Pull request by joostinyi**
- Implements Package patching support for live reload server
- Scopes this support to addition, update, and removal of packages after gather step since we cannot can't compute the patch between shadowed and normal truss handles. (i.e. patch is not calculated when adding external packages for the first time, or removing all external packages)

### CODE REVIEW

File: truss/patch/calc_patch.py [hunk 1]
```diff
@@ -100,19 +100,28 @@ def _under_unsupported_patch_dir(path: str) -> bool:
             # Don't support removal of config file
             logger.info(f"Patching not supported for removing {path}")
```
[the remark this answers is not in the document]
Comment (joostinyi) [reply truss/patch/calc_patch.py:101]: We don't want these logs to get handled? https://github.com/basetenlabs/truss/blob/893d5df24c655ccb1f3b69e60e52bd9609a9a61b/truss/templates/control/control/helpers/logging.py#L6-L7

File: truss/patch/calc_patch.py [hunk 2]
```diff
@@ -352,20 +373,13 @@ def _mk_config_patch(action: Action, config: dict) -> Patch:
     )
 
 
-def _mk_data_patch(action: Action, item: dict, path: str) -> Patch:
+def _mk_data_patch(action: Action, item: str, path: str) -> Patch:
```
[the remark this answers is not in the document]
Comment (joostinyi) [reply truss/patch/calc_patch.py:376]: I introduced this boilerplate a couple changes back – it isn't being used anywhere though. Let me add a comment to address this confusion

File: truss/patch/calc_patch.py [hunk 3]
```diff
@@ -129,10 +138,22 @@ def _under_unsupported_patch_dir(path: str) -> bool:
                 yaml.safe_load(previous_truss_signature.config)
             )
             config_patches = calc_config_patches(prev_config, new_config)
-            if config_patches is None:
-                logger.info(f"Unable to patch update to {path}")
-                return None
             patches.extend(config_patches)
+        elif path.startswith(bundled_packages_path):
```
[the remark this answers is not in the document]
Comment (joostinyi) [reply truss/patch/calc_patch.py:142]: I thought the `startswith` was more readable but happy to change them to use the helper

**Review by bolasim — APPROVED** [review]
Feel free to merge when you think its ready

File: truss/templates/control/control/helpers/truss_patch/model_code_patch_applier.py [hunk 4]
```diff
@@ -32,8 +34,15 @@ def apply_model_code_patch(
     elif action == Action.REMOVE:
         if not filepath.exists():
             logger.warning(f"Could not delete file {filepath}: not found.")
-        else:
+        elif filepath.is_file():
             logger.info(f"Deleting file {filepath}")
             filepath.unlink()
+            try:
+                os.removedirs(filepath.parent)
```
[the remark this answers is not in the document]
Comment (joostinyi) [reply truss/templates/control/control/helpers/truss_patch/model_code_patch_applier.py:41]: Wanted to attempt to remove parent directories recursively if the deletion of the file here makes the directories empty. Let me add a comment to clarify this

File: truss/templates/control/control/helpers/truss_patch/model_code_patch_applier.py [hunk 5]
```diff
@@ -32,8 +34,15 @@ def apply_model_code_patch(
...
+            except OSError as e:
+                if e.errno == 39:  # Directory not empty
```
[the remark this answers is not in the document]
Comment (joostinyi) [reply truss/templates/control/control/helpers/truss_patch/model_code_patch_applier.py:43]: exception is raised when we attempt to remove a non-empty directory, which we want to catch without raising.

**Pull request merged**